crates/store/src/db/sql/mod.rs

```diff
@@ -190,6 +193,317 @@ pub fn select_accounts_by_ids(
+        let key = row.get_ref(2)?.as_blob()?;
+        let key = Digest::read_from_bytes(key)?;
+        let value = row.get_ref(3)?.as_blob()?;
+        let value = Word::read_from_bytes(value)?;
+
+        match find_account(&mut accounts, account_id)?.get_mut().storage.entry(slot) {
+            Entry::Vacant(entry) => {
+                entry.insert(StorageSlot::Map(
+                    StorageMap::with_entries([(key, value)])
+                        .map_err(|err| DatabaseError::DataCorrupted(err.to_string()))?,
+                ));
+            },
+            Entry::Occupied(mut entry) => match entry.get_mut() {
+                StorageSlot::Value(_) => {
+                    return Err(DatabaseError::DataCorrupted(format!(
+                        "Conflicting storage slot: {slot}, expected map, but actually \
+                                    value"
+                    )))
+                },
+                StorageSlot::Map(map) => {
+                    map.insert(key, value);
+                },
+            },
+        }
+    }
+
+    let mut rows =
+        select_fungible_asset_deltas_stmt.query(params![Rc::clone(&account_ids), block_number])?;
+    while let Some(row) = rows.next()? {
+        let account_id: AccountId = row.get(0)?;
+        let faucet_id: AccountId = row.get(1)?;
+        let amount: u64 = row.get(2)?;
+        find_account(&mut accounts, account_id)?.get_mut().assets.push(Asset::Fungible(
+            FungibleAsset::new(
+                faucet_id
+                    .try_into()
+                    .map_err(|err: AccountError| DatabaseError::DataCorrupted(err.to_string()))?,
+                amount,
+            )
+            .map_err(|err| DatabaseError::DataCorrupted(err.to_string()))?,
+        ));
+    }
+
+    let mut rows = select_latest_non_fungible_assets_stmt
+        .query(params![Rc::clone(&account_ids), block_number])?;
+    while let Some(row) = rows.next()? {
+        let account_id: AccountId = row.get(0)?;
+        let vault_key = row.get_ref(1)?.as_blob()?;
+        let vault_key = Word::read_from_bytes(vault_key)?;
+        let asset = NonFungibleAsset::try_from(vault_key)
+            .map_err(|err| DatabaseError::DataCorrupted(err.to_string()))?;
+        find_account(&mut accounts, account_id)?
+            .get_mut()
+            .assets
+            .push(Asset::NonFungible(asset));
+    }
+
+    // Converting gathered data to vector of `Account` structures.
+    let mut result = Vec::with_capacity(account_ids.len());
+    for (account_id, record) in accounts {
+        let slots: Vec<_> = record
+            .storage
+            .into_iter()
+            .enumerate()
+            .map(|(expected_slot, (slot, value))| {
+                if expected_slot != slot as usize {
+                    return Err(DatabaseError::DataCorrupted(format!(
+                        "Missing value for storage slot {expected_slot}, got {slot}"
+                    )));
```

Comment:
Is there value in expanding `DataCorrupted` with variants for each possible corruption? Probably not 🤔

Reply:
No, I don't think we will process any data corruption cases differently from any other. :)